internal/certgen/certgen.go

```diff
@@ -113,37 +158,84 @@ func NewContourCerts() *ContourCerts {
-	return dumpFile(outputDir+"/"+cd.Filename, cd.Data)
+// writeCACertYAML is a helper function to write out just the CA's certificate
+// as a Kubernetes Secret. Required so that we don't give access to the full
+// keypair to consumers like Contour and Envoy.
+func (cc *ContourCerts) writeCACertYAML(outputDir, namespace string) error {
+	secret := newCertOnlySecret(cc.CACertSecretName, namespace, cc.CA.Cert.Filename, cc.CA.Cert.Data)
+	return writeSecret(outputDir+"/"+cc.CACertSecretName+".yaml", secret)
+}
 
+// writeCACertKube writes out just the CA's certificate as a Kubernetes Secret.
+// Required so that we don't give access to the full keypair to consumers like Contour and Envoy.
+func (cc *ContourCerts) writeCACertKube(namespace string, client *kubernetes.Clientset) error {
+	secret := newCertOnlySecret(cc.CACertSecretName, namespace, cc.CA.Cert.Filename, cc.CA.Cert.Data)
+	_, err := client.CoreV1().Secrets(namespace).Create(secret)
+	if err != nil {
+		return err
+	}
+	fmt.Printf("secret/%s created\n", cc.CACertSecretName)
+	return nil
 }
 
-// writeCertPEMs writes out all the PEMs for all the certs, using
-// the stored filenames.
-// TODO(youngnick) we should be able to use a similar pattern for the
-// secrets, hopefully.
-func (cc *ContourCerts) writeCertPEMs(outputDir string) error {
-	err := cc.CA.writePEMs(outputDir)
+// writeSecretYAMLs writes out all the secret YAMLs to `outputDir`, putting
+// the generated Secrets into the Kube `namespace`
+func (cc *ContourCerts) writeSecretYAMLs(outputDir, namespace string) error {
+	// First, write out just the CA Cert secret.
+	err := cc.writeCACertYAML(outputDir, namespace)
 	if err != nil {
 		return err
 	}
-	err = cc.Contour.writePEMs(outputDir)
+	// Next, write out the full CA keypair
+	err = cc.CA.writeTLSYAML(outputDir, namespace)
 	if err != nil {
 		return err
 	}
-	return cc.Envoy.writePEMs(outputDir)
+	// Next, Contour's keypair
+	err = cc.Contour.writeTLSYAML(outputDir, namespace)
+	if err != nil {
+		return err
+	}
+	return cc.Envoy.writeTLSYAML(outputDir, namespace)
+}
 
+// writeToKube writes all the secrets directly to Kubernetes, in the given
+// `namespace`.
+func (cc *ContourCerts) writeToKube(namespace string, client *kubernetes.Clientset) error {
+	// First, write out just the CA Cert secret.
+	err := cc.writeCACertKube(namespace, client)
+	if err != nil {
+		return err
+	}
+	// Next, write out the full CA keypair
+	err = cc.CA.writeTLSKube(namespace, client)
+	if err != nil {
+		return err
+	}
+	// Next, Contour's keypair
+	err = cc.Contour.writeTLSKube(namespace, client)
+	if err != nil {
+		return err
+	}
+	return cc.Envoy.writeTLSKube(namespace, client)
 }
 
-// GenerateCerts performs the actual cert generation steps and then returns the certs for the output functions.
+// GenerateCerts performs the actual cert generation steps and then returns the certs for the output function.
 func GenerateCerts(certConfig *Config) (*ContourCerts, error) {
```

Comment:
I'm pretty sure this needs to move to cmd/contour. certConfig is 100% the cli flags we pass into `contour gencerts` so its probably better than certConfig never leaks out of cmd/contour.